### DISCUSSION TITLE
update public profiles whenever auth.users metadata updated, inserted, deleted

### DISCUSSION THREAD

**Discussion by amrajat** [supabase/supabase#20330] [Questions] [answered]
I created a public profiles table using the user management starter template and modified its columns and functions, but it's not working as it should be.

supabase.auth.signInWithOtp function returns this.

Error:  ⨯ Error: AuthApiError: duplicate key value violates unique constraint "profiles_pkey"

Below are my triggers:

on_auth_user_metadata_update | users | handle_new_user | AFTER UPDATE |  

Below is my function:

handle_new_user | - | trigger | Definer

begin
  insert into public.profiles (id, full_name, avatar_url, website, company, position, x_username, permission)
  values (new.id, 
          new.raw_user_meta_data ->> 'full_name', 
          new.raw_user_meta_data ->> 'avatar_url', 
          new.raw_user_meta_data ->> 'website', 
          new.raw_user_meta_data ->> 'company', 
          new.raw_user_meta_data ->> 'position', 
          new.raw_user_meta_data ->> 'x_username', 
          new.raw_user_meta_data['permission']::boolean);
  return new;
end;

**Comment by GaryAustin1** [comment, marked as answer]
You are triggering on update to auth.users, but doing an insert to your profiles table.  Update will continue to be called many times and the 2nd time would error for you.

You need an insert trigger function and an update trigger function that does updates you your profile table, OR one function can do both but you have to have an insert part and an update part.  If you also want delete then that is yet a separate function or set of code.

https://www.postgresql.org/docs/current/plpgsql-trigger.html
This shows TG_OP if you want to combine into one function, but it might be easier for you just to have 3 functions if just learning SQL.

**Reply by amrajat** [reply]
Thank you for explaining, I dropped all the triggers and functions and created them separately.
